### PR TITLE
support gemm fp8/fp16 mixed input

### PR DESCRIPTION
**Pull request by scxiao**

### CODE REVIEW

File: python/test/unit/language/test_core_amd.py [hunk 1]
```diff
@@ -1043,11 +1043,146 @@ def copy_kernel(input_ptr, output_ptr, n_elements, BLOCK_SIZE: tl.constexpr):
...
+
+    @triton.jit
+    def copy_kernel(input_ptr, output_ptr, n_elements, BLOCK_SIZE: tl.constexpr):
+        offsets = tl.program_id(axis=0) * BLOCK_SIZE + tl.arange(0, BLOCK_SIZE)
+        mask = offsets < n_elements
+        input = tl.load(input_ptr + offsets, mask=mask)
+        output = input
+        tl.store(output_ptr + offsets, output, mask=mask)
+
+    @triton.autotune(
```
Comment (zhanglx13) [inline python/test/unit/language/test_core_amd.py:1077]: Are we asked to measure performance for this mixed gemm?
If not, we can remove this autotune, which leaves the impression that we are tuning to get better perf.

File: lib/Conversion/TritonGPUToLLVM/TypeConverter.cpp [hunk 2]
```diff
@@ -79,6 +79,35 @@ Value TritonGPUToLLVMTypeConverter::packLLElements(
...
+  if (!(encoding && encoding.getParent().isa<MfmaEncodingAttr>())) {
+    return inValues;
+  }
+
+  SmallVector<Value> result;
+  auto structType = this->convertType(srcTy).dyn_cast<LLVM::LLVMStructType>();
+  auto elementTypes = structType.getBody();
+  assert(elementTypes.size() > 0);
+  auto vecTy = elementTypes[0];
+  unsigned size = 4;
```
Comment (zhanglx13) [inline lib/Conversion/TritonGPUToLLVM/TypeConverter.cpp:98]: is this hard-coded due to the assumption of mfma_32x32x8f16?

File: lib/Conversion/TritonGPUToLLVM/Utility.cpp [hunk 3]
```diff
@@ -289,4 +289,10 @@ Value addStringToModule(Location loc, ConversionPatternRewriter &rewriter,
 }
 
 } // namespace LLVM
+
+bool isF8(Type eType) {
+  return eType.isFloat8E5M2FNUZ() or eType.isFloat8E4M3FNUZ() or eType.isFloat8E5M2() or eType.isFloat8E5M2FNUZ();
```
Comment (zhanglx13) [inline lib/Conversion/TritonGPUToLLVM/Utility.cpp:294]: This is probably to long to fit into a single line. Can you run clang to fix the format for the files this PR has touched?

File: lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMFMA.cpp [hunk 4]
```diff
@@ -439,6 +445,9 @@ Value loadA(ConversionPatternRewriter &rewriter, Location loc, Value thread,
       std::max<int>(mfmaInstrM * mfmaInstrK / iWaveSize /*wave size*/, 1);
   unsigned int maxNumWarps = shape[0] / mfmaInstrM;
   int warpsPerGroupM = std::min(warpsPerCTA[0], maxNumWarps);
+  if (isF8(aElemTy)) {
+    aElemTy = i8_ty;
+  }
```
Comment (binarman) [inline lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMFMA.cpp:450]: Optional:
You can use TypeConverter for this:
```suggestion
    aElemTy = typeConverter->convertType(aElemTy);
```
Plus I want to refactor code on line 471 and 492, you can do it in this PR, if you want.

File: lib/Conversion/TritonGPUToLLVM/Utility.cpp (hunk 3 above)
Comment (binarman) [reply lib/Conversion/TritonGPUToLLVM/Utility.cpp:294]: @scxiao 
You can use `git-clang-format <commit hash>`  to reformat code to comply with `.clang-format` config of our project.

This toll will not change code which was changed before given commit, so you can reformat only your code.

**Review by binarman — APPROVED** [review]
LGTM
P.s. I do not have rights to approve by this account, so this PR will not be marked as approved

File: python/test/unit/language/test_core_amd.py (hunk 1 above)
Comment (scxiao) [reply python/test/unit/language/test_core_amd.py]: Thanks. Removed.

File: lib/Conversion/TritonGPUToLLVM/TypeConverter.cpp (hunk 2 above)
Comment (scxiao) [reply lib/Conversion/TritonGPUToLLVM/TypeConverter.cpp]: Change to call `unsigned size = vecTy.getNumElements();` for the vec size.

File: python/triton/language/semantic.py [hunk 5]
```diff
@@ -1301,13 +1301,19 @@ def dot(lhs: tl.tensor,
...
     assert lhs.shape[1].value == rhs.shape[0].value, f"First input shape ({lhs.shape}) and second input shape {rhs.shape} are not compatible for matmul (second index of first shape ({lhs.shape[1].value}) must be equal to first index of second shape ({rhs.shape[0].value})"
     assert lhs.shape[0].value >= 16 and lhs.shape[1].value >= 16 \
         and rhs.shape[1].value >= 16,\
         f"All values in both first input shape ({lhs.shape}) and second input shape ({rhs.shape}) must be >= 16!"
+
+    if lhs.type.scalar.is_fp8():
+        lhs = cast(lhs, tl.float16, builder)
+    elif rhs.type.scalar.is_fp8():
+        rhs = cast(rhs, tl.float16, builder)
+
```
[the remark this answers is not in the document]
Comment (scxiao) [reply python/triton/language/semantic.py:1316]: OK, we can take a look at this after IFU.

File: lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMFMA.cpp (hunk 4 above)
Comment (scxiao) [reply lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMFMA.cpp]: Thanks. Changed all places.

**Comment by scxiao** [comment]
Hi @zhanglx13 Do you have other comments for this PR? If not, could you please help approve it? Thanks

**Comment by zhanglx13** [comment]
@scxiao Let me run some benchmarks to make sure there is no regression on gemm perfs.

**Review by zhanglx13 — APPROVED** [review]
LGTM and no regression on GEMM perfs

**Comment by zhanglx13** [comment]
@scxiao There are 19 commits in this PR. Do you want to keep some or all of them? Is it OK to squash and merge?

**Comment by scxiao** [comment]
> @scxiao There are 19 commits in this PR. Do you want to keep some or all of them? Is it OK to squash and merge?

Yes, I will squash them. No need to keep those individual commits

**Pull request merged**